File: src/main/kotlin/com/feakin/intellij/structure/FeakinStructureViewElement.java
package com.feakin.intellij.structure;

import com.feakin.intellij.FeakinFile;
import com.feakin.intellij.psi.FeakinContextDeclaration;
import com.feakin.intellij.psi.FeakinContextMapDeclaration;
import com.feakin.intellij.psi.FeakinContextMapName;
import com.feakin.intellij.psi.FeakinContextName;
import com.feakin.intellij.psi.impl.FeakinContextMapDeclarationImpl;
import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.util.treeView.smartTree.SortableTreeElement;
import com.intellij.ide.util.treeView.smartTree.TreeElement;
import com.intellij.navigation.ItemPresentation;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.psi.NavigatablePsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class FeakinStructureViewElement implements StructureViewTreeElement, SortableTreeElement  {
    private final NavigatablePsiElement myElement;

    public FeakinStructureViewElement(NavigatablePsiElement element) {
        this.myElement = element;
    }

    @Override
    public Object getValue() {
        return myElement;
    }

    @Override
    public void navigate(boolean requestFocus) {
        myElement.navigate(requestFocus);
    }

    @Override
    public boolean canNavigate() {
        return myElement.canNavigate();
    }

    @Override
    public boolean canNavigateToSource() {
        return myElement.canNavigateToSource();
    }

    @NotNull
    @Override
    public String getAlphaSortKey() {
        String name = myElement.getName();
        return name != null ? name : "";
    }

    @NotNull
    @Override
    public ItemPresentation getPresentation() {
        PresentationData struct = new PresentationData();
        struct.setIcon(null);
        struct.setLocationString(null);

        if (myElement instanceof FeakinContextMapDeclarationImpl) {
            FeakinContextMapName structNameDeclaration = ((FeakinContextMapDeclarationImpl) myElement).getContextMapName();

            struct.setAttributesKey(DefaultLanguageHighlighterColors.KEYWORD);
            struct.setPresentableText(structNameDeclaration.getText());
            return struct;
        } else if (myElement instanceof FeakinContextDeclaration) {
            FeakinContextDeclaration myElement = (FeakinContextDeclaration) this.myElement;

            FeakinContextName structNameDeclaration = myElement.getContextName();

            struct.setAttributesKey(DefaultLanguageHighlighterColors.KEYWORD);
            struct.setPresentableText(structNameDeclaration.getText());
            return struct;

        }

        ItemPresentation presentation = myElement.getPresentation();
        return presentation != null ? presentation : new PresentationData();
    }

    @NotNull
    @Override
    public TreeElement[] getChildren() {
        if (myElement instanceof FeakinFile) {
            List<FeakinContextMapDeclaration> properties = PsiTreeUtil.getChildrenOfTypeAsList(myElement, FeakinContextMapDeclaration.class);
            List<TreeElement> treeElements = new ArrayList<>(properties.size());
            List<FeakinContextDeclaration> methodProperties = PsiTreeUtil.getChildrenOfTypeAsList(myElement, FeakinContextDeclaration.class);

            for (FeakinContextMapDeclaration property : properties) {
                treeElements.add(new FeakinStructureViewElement(property));
            }

            for (FeakinContextDeclaration property : methodProperties) {
                treeElements.add(new FeakinStructureViewElement(property));
            }

            return treeElements.toArray(new TreeElement[0]);
        }
        return EMPTY_ARRAY;
    }

}

File: src/main/kotlin/com/feakin/intellij/structure/FeakinStructureViewElement.kt
package com.feakin.intellij.structure

import com.feakin.intellij.FkFile
import com.feakin.intellij.psi.FeakinContextDeclaration
import com.feakin.intellij.psi.FeakinContextMapDeclaration
import com.feakin.intellij.psi.FeakinNamedElement
import com.feakin.intellij.psi.impl.FeakinContextMapDeclarationImpl
import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.util.treeView.TreeAnchorizer
import com.intellij.ide.util.treeView.smartTree.SortableTreeElement
import com.intellij.ide.util.treeView.smartTree.TreeElement
import com.intellij.navigation.ItemPresentation
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.pom.Navigatable
import com.intellij.psi.NavigatablePsiElement
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil

class FeakinStructureViewElement(private val myElement: NavigatablePsiElement) : StructureViewTreeElement,
    SortableTreeElement {

    private val psiAnchor = TreeAnchorizer.getService().createAnchor(myElement)

    private val psi: PsiElement? get() = TreeAnchorizer.getService().retrieveElement(psiAnchor) as? PsiElement

    override fun navigate(requestFocus: Boolean) {
        (psi as? Navigatable)?.navigate(requestFocus)
    }

    override fun canNavigate(): Boolean = (psi as? Navigatable)?.canNavigate() == true

    override fun canNavigateToSource(): Boolean = (psi as? Navigatable)?.canNavigateToSource() == true

    override fun getValue(): PsiElement? = psi

    override fun getAlphaSortKey(): String {
        val name = myElement.name
        return name ?: ""
    }

    override fun getPresentation(): ItemPresentation {
        val struct = PresentationData()
        struct.setIcon(null)
        struct.locationString = null
        if (myElement is FeakinContextMapDeclaration) {
            val structNameDeclaration = myElement.contextMapName
            struct.setAttributesKey(DefaultLanguageHighlighterColors.KEYWORD)
            struct.presentableText = structNameDeclaration.text
            return struct
        } else if (myElement is FeakinContextDeclaration) {
            val structNameDeclaration = myElement.contextName
            struct.setAttributesKey(DefaultLanguageHighlighterColors.KEYWORD)
            struct.presentableText = structNameDeclaration.text
            return struct
        }
        val presentation = myElement.presentation
        return presentation ?: PresentationData()
    }

    override fun getChildren(): Array<TreeElement> {
        if (myElement is FkFile) {
            val properties = PsiTreeUtil.getChildrenOfTypeAsList(myElement, FeakinContextMapDeclaration::class.java)
            val treeElements: MutableList<TreeElement> = ArrayList(properties.size)
            val methodProperties = PsiTreeUtil.getChildrenOfTypeAsList(myElement, FeakinContextDeclaration::class.java)
            for (property in properties) {
                treeElements.add(FeakinStructureViewElement(property))
            }
            for (property in methodProperties) {
                treeElements.add(FeakinStructureViewElement(property))
            }
            return treeElements.toTypedArray()
        }

        return arrayOf()
    }
}